packages/gatsby/src/redux/reducers/queries.ts

```diff
@@ -0,0 +1,209 @@
+  return {
+    byNode: new Map<NodeId, Set<QueryId>>(),
+    byConnection: new Map<ConnectionName, Set<QueryId>>(),
+    trackedQueries: new Map<QueryId, IQueryState>(),
+    trackedComponents: new Map<ComponentPath, IComponentState>(),
+  }
+}
+
+const initialQueryState = (): IQueryState => {
+  return {
+    dirty: -1, // unknown, must be set right after init
+  }
+}
+
+const initialComponentState = (): IComponentState => {
+  return {
+    componentPath: ``,
+    query: ``,
+    pages: new Set<QueryId>(),
+    errors: 0,
+    // TODO: staticQueries: new Set<QueryId>()
+  }
+}
+
+/**
+ * Tracks query dirtiness. Dirty queries are queries that:
+ *
+ * - depend on nodes or node collections (via `actions.createPageDependency`) that have changed.
+ * - have been recently extracted (or their query text has changed)
+ * - belong to newly created pages (or pages with modified context)
+ *
+ * Dirty queries must be re-ran.
+ */
+export function queriesReducer(
+  state: IGatsbyState["queries"] = initialState(),
+  action: ActionsUnion
+): IGatsbyState["queries"] {
+  switch (action.type) {
+    case `DELETE_CACHE`:
+      return initialState()
+
+    case `CREATE_PAGE`: {
+      const { path, componentPath } = action.payload
+      if (!state.trackedQueries.has(path) || action.contextModified) {
+        const query = registerQuery(state, path)
+        query.dirty = setFlag(query.dirty, FLAG_DIRTY_PAGE)
+      }
+      registerComponent(state, componentPath).pages.add(path)
+      return state
+    }
+    case `DELETE_PAGE`: {
+      const { path, componentPath } = action.payload
+      const component = state.trackedComponents.get(componentPath)
+      if (component) {
+        component.pages.delete(path)
+      }
+      state.trackedQueries.delete(path)
+      return state
+    }
+    case `QUERY_EXTRACTED`: {
+      // Note: this action is called even in case of
+      // extraction error or missing query (with query === ``)
+      // TODO: use hash instead of a query text
+      const { componentPath, query } = action.payload
+      const component = registerComponent(state, componentPath)
+      if (component.query !== query) {
+        // Invalidate all pages associated with a component when query text changes
+        component.pages.forEach(queryId => {
+          const query = state.trackedQueries.get(queryId)
+          if (query) {
+            query.dirty = setFlag(query.dirty, FLAG_DIRTY_TEXT)
+          }
+        })
+        component.query = query
+      }
+      return state
+    }
+    case `QUERY_EXTRACTION_GRAPHQL_ERROR`:
+    case `QUERY_EXTRACTION_BABEL_ERROR`:
+    case `QUERY_EXTRACTION_BABEL_SUCCESS`: {
+      const { componentPath } = action.payload
+      const component = registerComponent(state, componentPath)
+      const set = action.type !== `QUERY_EXTRACTION_BABEL_SUCCESS`
+      component.errors = setFlag(component.errors, FLAG_ERROR_BABEL, set)
+      return state
+    }
```

Comment:
Marking a component that failed to parse (babel or GraphQL). We won't run queries for components with errors (at least until they recover, i.e. user fixes the error).